subsys/net/lib/socks/socks.c

```diff
@@ -0,0 +1,157 @@
+
+/**@brief Connects to destination through a SOCKS5 proxy server.
+ *
+ * @param[in] proxy Address of the proxy server.
+ * @param[in] destination Address of the destination server.
+ *
+ * @retval File descriptor of the opened connection or an error code if it was
+ *         unsuccessful.
+ */
+int socks5_client_tcp_connect(const struct sockaddr *proxy,
+			      const struct sockaddr *destination)
+{
+	const struct sockaddr_in *p = (struct sockaddr_in *) proxy;
+
+	struct socks5_method_request mthd_req;
+	struct socks5_method_response mthd_rep;
+	struct socks5_command_request cmd_req;
+	struct socks5_command_response cmd_rep;
+	int size;
+	int ret;
+	int fd;
+
+	fd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+
+	if (fd < 0) {
+		return fd;
+	}
+
+	ret = connect(fd, proxy, sizeof(struct sockaddr_in));
+
+	if (ret < 0) {
+		LOG_ERR("Unable to connect to the proxy server");
+		(void)close(fd);
+		return -errno;
+	}
+
+	/* Negotiate authentication method */
+	mthd_req.r.ver = SOCKS5_PKT_MAGIC;
+
+	/* We only support NOAUTH at the moment */
+	mthd_req.r.nmethods = 1;
+	mthd_req.methods[0] = SOCKS5_AUTH_METHOD_NOAUTH;
+
+	/* size + 1 because just one method is supported */
+	size = sizeof(struct socks5_method_request_common) + 1;
+	ret = send(fd, &mthd_req, size, 0);
+
+	if (ret != size) {
+		(void)close(fd);
+		LOG_ERR("Could not send negotiation packet");
+		return -EINVAL;
+	}
+
+	ret = recv(fd, &mthd_rep, sizeof(mthd_rep), 0);
+
+	if (ret != sizeof(struct socks5_method_response)) {
+		LOG_ERR("Could not receive negotiation response");
+		(void)close(fd);
+		return -EINVAL;
+	}
+
+	if (mthd_rep.ver != SOCKS5_PKT_MAGIC) {
+		LOG_ERR("Invalid negotiation response magic");
+		(void)close(fd);
+		return -EINVAL;
+	}
+
+	if (mthd_rep.method != SOCKS5_AUTH_METHOD_NOAUTH) {
+		LOG_ERR("Invalid negotiation response");
+		(void)close(fd);
+		return -ENOTSUP;
+	}
+
+	/* Negotiation complete - now connect to destination */
+	cmd_req.r.ver = SOCKS5_PKT_MAGIC;
+	cmd_req.r.cmd = SOCKS5_CMD_CONNECT;
+	cmd_req.r.rsv = SOCKS5_PKT_RSV;
+
+	if (p->sin_family == AF_INET) {
+		const struct sockaddr_in *d4 =
+			(struct sockaddr_in *) destination;
+
+		cmd_req.r.atyp = SOCKS5_ATYP_IPV4;
+
+		memcpy(&cmd_req.ipv4_addr.addr,
+		       (u8_t *)&d4->sin_addr,
+		       sizeof(cmd_req.ipv4_addr.addr));
+
+		cmd_req.ipv4_addr.port = d4->sin_port;
+
+		size = sizeof(struct socks5_command_request_common)
+			+ sizeof(struct socks5_ipv4_addr);
+	} else if (p->sin_family == AF_INET6) {
+		const struct sockaddr_in6 *d6 =
+			(struct sockaddr_in6 *) destination;
+
+		cmd_req.r.atyp = SOCKS5_ATYP_IPV6;
+
+		memcpy(&cmd_req.ipv6_addr.addr,
+		       (u8_t *)&d6->sin6_addr,
+		       sizeof(cmd_req.ipv6_addr.addr));
+
+		cmd_req.ipv4_addr.port = d6->sin6_port;
+
+		size = sizeof(struct socks5_command_request_common)
+			+ sizeof(struct socks5_ipv6_addr);
+	}
+
+	ret = send(fd, &cmd_req, size, 0);
+
+	if (ret != size) {
+		LOG_ERR("Could not send CONNECT command");
```

Comment:
That won't work, sorry. It's not error to have only partial data to be sent. You need to retry with the rest. (Or maybe we need to implement MSG_WAITALL after all for convenience, but looking thru man, it's defined only for recv(), not for send()! ).

Reply:
> Or maybe we need to implement MSG_WAITALL after all for convenience, but looking thru man, it's defined only for recv(), not for send()!

This is now ticketed as https://github.com/zephyrproject-rtos/zephyr/issues/13091

Reply:
I've added the helper functions based on how it was done in other places, however the current implementation is fully blocking. I guess that is ok in this case?

Reply:
> however the current implementation is fully blocking.

My response would be: "How else could it be? Writing non-blocking applications is rather complicated." But, you want to use it for mqtt lib, which went thru the trouble of implementing it in non-blocking way. Please work with @rlubos to ensure that it all will work as it should. I would think it should: you perform initial handshake in a blocking manner, then mqtt lib will just switch sock5 socket into non-blocking mode.